pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md

```diff
@@ -0,0 +1,303 @@
+
+With this scenario in mind we set out to find solutions and all solutions we could think of required haggling
+with Kubernetes.
+
+## Extending Kubernetes
+
+There are several ways to extend Kubernetes. All functionality in Kubernetes is build upon very nice and clean
+public APIs, or to say it with other words: There is no private API magic hidden somewhere. So lets look at
+two ways on how to extend Kubernetes.
+
+Possible ways to go forward:
+- Adding a scheduler extender
+- Creating an operator/controller
+
+There are more ways to extend Kubernetes, but these two ways will be the one we shall look at. Just for completeness
+you can as well also add another scheduler or change kubernetes itself. However these possibilities have some serious
+downsides.
+
+### Adding a scheduler extender
+The kubernetes scheduler checks for certain requirements before it schedules a pod onto a node. Some of these requirements
+are hard requirements, like cpu, memory and number of pods. Other requirements are more soft, like if the pods are allowed to
+be packed together or in which AZ they are going to run. All of those requirements are collected and points given to each
+node on how good they meet the requirements. The node that fits best, gets chosen.
+
+All of that are things the scheduler will do for you automatically, however it is also possible to give the cluster a 
+`KubeSchedulerConfiguration` object that will tell the scheduler to also reach out to a service for additional point scoring.
+The SchedulerConfiguration is a JSON file and for further explanations, please have a look at this 
+excellent [blog post](https://developer.ibm.com/articles/creating-a-custom-kube-scheduler/).
+
+In our case, we could have written a service that checks which IP Addresses are assigned to the Nodes and moved the Pods
+onto those nodes. This would have required us to make sure that those nodes had all needed IP-Addresses all the time.
+That sounded not very enticing when doing cluster upgrades as it would have needed to be at least a semi-manual process.
+So we decided against this approach.
+
+### Creating an operator
+An Operator/Controller on the other hand is a component observing resources and then try to create the declared resources.
+The difference between a controller and an operator is basically that an operator is handling the lifecycle of an
+application, whereas a controller may control specific resources that are not associated with a specific application.
+They both use the controller pattern though and both can be implemented with the same toolset, so for simplicity sake in the
+context of this article, we will consider them to be equal.
+
+Operators usually consists at least out of a **CRD(CustomResourceDefinition)**, an **event listener** and a **reconciliation loop**.
+The *CRD* defines the object that shall be created, e.g. a Redis deployment. The whole description on what this deployment
```

Comment:
Maybe point out that the CRD is creating a new type in Kubernetes and effectively implements a domain-specific language for the operator.

Reply:
You could also mention here that the new type is called a *kind* later